### PR TITLE
Integrate questionary

### PR DESCRIPTION
**Pull request by akkuldn**
This is a new feature that allows the user to run the tests interactively via the terminal.
__Steps to run:__
1) If you're using Gitbash for windows please make sure to set your bash alias by adding the following command to your `bash_rc` file `alias python='winpty python.exe'`
2) Install the requirements using `pip install -r reqquirements.txt`
3) Commands to run the test : 
`python -m pytest tests/test_example_form.py --interactivemode_flag Y`
`python -m pytest tests/test_mobile_bitcoin_price.py --interactivemode_flag Y`
` python -m pytest tests/test_api_example.py --interactivemode_flag Y`

Changes made:
---
- A new feature to run the tests interactive using `--interactivemode_flag`
- A new pytest fixture `test_api_obj` in contest file for API tests
- Importing the API URL from a conf file

### CODE REVIEW

File: conftest.py
Comment (qxf2) [inline conftest.py:93]: Why does the new argument not have a default?

File: conftest.py
Comment (qxf2) [inline conftest.py:22]: The whiteline spacing here is off. Why are there empty lines after the if?

File: endpoints/API_Player.py
Comment (qxf2) [inline endpoints/API_Player.py:14]: How are commented out lines of code slipping through to a PR?

File: utils/interactive_mode.py
Comment (qxf2) [inline utils/interactive_mode.py:1]: @akkuldn - I feel like this file will be extremely hard to maintain. A few suggestions:
a) Break your methods down better. Typically, methods should not be more than 10 lines. You should break the code down better.
b) When you intersperse code and strings, the strings should become configurable. If you are using the same strings in two or more places, you should be making it a variable. that way, when the hardcoded string needs to change, the QA changes them in only one spot.
c) Can you please add some unit tests for these options? Include them in the PR to help your reviewer. You can delete the tests once they approve your PR.

**Review by qxf2 — CHANGES_REQUESTED** [review]
@akkuldn - I got a little time to check out your work. This is a very good addition to our framework. 

I have left some comments based on ~5-minutes of code review. Please consider them along with the feedback you get from your other reviewers.

**Comment by PreedhiVivek** [comment]
@akkuldn 
An interesting and useful feature add-on to our framework. Nice Akkul!
My observations:
- Errors out while selecting the firefox version
browser_version_error

- Do you want to run the test with a default set of options? No -> What would you like to change? Exit -> Errors out. Would be good to go back to the default options question on Exit. 
error_on_exit

- ask_questions_<>() functions will have to be unit tested well, while modifying different default options and ensuring expected behaviour.

- In Readme.md '(This will give the user a menu of different choices allowing the user to pick their desired configuration to run the test)' can be rephrased to (This option will allow the user to pick the desired configuration to run the test, from the menu displayed)' and typo `aliash`.  Is that __Note__ even required? Isn't that understood and holds good for running the test suit with any other option as well.  You can probably make it a common note under that section?

- conftest.py Watch out for line spacing. Could see many empty lines and very long lines of code. I suggest, running your modified files against pylint and radon. You could fix the ones that belong to your change, if all is too much or parked as a separate task.

- API_Player.py remove commented out code line # 14, 26, 27  

- Interactive_mode.py, functions such as ask_questions_gui() and ask_questions_mobile() has be broken down. They are too long. Also, ensure this new module is pylint and radon compliant.

- Commits will have to be squashed before merge.  

- I will be running the API tests and will update if I find any comments.

File: Readme.md
Comment (avinash010) [inline Readme.md:154]: There are empty lines here too.

File: Readme.md
Comment (avinash010) [inline Readme.md:153]: Shouldn't this be named interactive_mode_flag as per the naming convention we follow?

File: conftest.py
Comment (akkuldn) [reply conftest.py]: Fixed

File: endpoints/API_Player.py
Comment (akkuldn) [reply endpoints/API_Player.py]: Fixed

File: utils/interactive_mode.py
Comment (akkuldn) [reply utils/interactive_mode.py:1]: a) Brokedown the methods making them smaller
b) Now having a single variable for most of the reused strings
c) Added some unit tests
Fixed

**Comment by akkuldn** [comment]
> @akkuldn
> An interesting and useful feature add-on to our framework. Nice Akkul!
> My observations:
> 
> * Errors out while selecting the firefox version, when I want to change default options:
>   browser_version_error
> * Do you want to run the test with a default set of options? No -> What would you like to change? Exit -> Errors out. Would be good to go back to the default options question on Exit.
>   error_on_exit
> * ask_questions_<>() functions will have to be unit tested well, while modifying different default options and ensuring expected behaviour.
> * In Readme.md '(This will give the user a menu of different choices allowing the user to pick their desired configuration to run the test)' can be rephrased to (This option will allow the user to pick the desired configuration to run the test, from the menu displayed)' and typo `aliash`.  Is that **Note** even required? Isn't that understood and holds good for running the test suit with any other option as well.  You can probably make it a common note under that section?
> * conftest.py Watch out for line spacing. Could see many empty lines and very long lines of code. I suggest, running your modified files against pylint and radon. You could fix the ones that belong to your change, if all is too much or parked as a separate task.
> * API_Player.py remove commented out code line # 14, 26, 27
> * Interactive_mode.py, functions such as ask_questions_gui() and ask_questions_mobile() has be broken down. They are too long. Also, ensure this new module is pylint and radon compliant.
> * Commits will have to be squashed before merge.
> * I will be running through the API test changes and will update if I find any comments.

---
Thanks for the feedback Preedhi,
1) Fixed
2) I added the exit option intentionally so that the user could exit out of the while loop incase he doesn't want to run the test
3) Added unit tests, you can find it inside the `tests` directory `test_interactive_mode.py`
4) Fixed
5) Fixed
6) Fixed
7) Fixed
8) Ill squash the commits once  I get the approval to merge

Please do let me know of any further comments

File: Readme.md
Comment (akkuldn) [reply Readme.md:156]: Fixed

File: Readme.md
Comment (akkuldn) [reply Readme.md]: Fixed

**Comment by PreedhiVivek** [comment]
@akkuldn 
My re-reviewing comments:
Readme file:
- Remove empty line # 152,  _Note:_ to NOTE: can be changed to be in sync with existing convention and would be good to have `interactive_mode` as the optional command line parameter in-place of `interactive_mode_flag`

conftest.py
- unwanted print line # 97

> > * Errors out while selecting the firefox version, when I want to change default options:
> >   browser_version_error
Test runs smoothly if I change browser to Firefox. However, when I tried Firefox+Windows+10 combination, observed that
Firefox browser launches and the cursor keeps blinking at the address bar. In the console,
`Errros out when I select Firefox, Windows, version 10
Exception when trying to run test: C:\Users\PreedhiVivek\code\POM\qxf2-page-object-model\conftest.py
Python says:Message: permission denied`
> > * ask_questions_<>() functions will have to be unit tested well, while modifying different default options and ensuring expected behaviour.
Unit tests should ensure all such combinations are covered. Please ensure that.

> > * Do you want to run the test with a default set of options? No -> What would you like to change? Exit -> Errors out. Would be good to go back to the default options question on Exit.
> >   error_on_exit
I understand the use-case Akkul. But the handling of this case should be looked into. It currently errors out when Exit is chosen from the menu. Maybe, catch the exception and log a good message.

> > * Interactive_mode.py, functions such as ask_questions_gui() and ask_questions_mobile() has be broken down. They are too long. Also, ensure this new module is pylint and radon compliant.
I do see that, you have broken down code well now. However, it has to be done a bit more. Pylint of this module is rated at 7.72/10 and radon run also points out few functions rated B/C which need to be addressed.

**Comment by akkuldn** [comment]
> @akkuldn
> My re-reviewing comments:
> Readme file:
> 
> * Remove empty line # 152,  `_Note:_` to NOTE: can be changed to be in sync with existing convention and would be good to have `interactive_mode` as the optional command line parameter in-place of `interactive_mode_flag`
> 
> conftest.py
> 
> * unwanted print line # 97
> 
> > > * Errors out while selecting the firefox version, when I want to change default options:
> > >   browser_version_error
> 
> Test runs smoothly if I change browser to Firefox. However, when I tried Firefox+Windows+10 combination, observed that
> Firefox browser launches and the cursor keeps blinking at the address bar. In the console,
> `Errros out when I select Firefox, Windows, version 10 Exception when trying to run test: C:\Users\PreedhiVivek\code\POM\qxf2-page-object-model\conftest.py Python says:Message: permission denied`
> 
> > > * ask_questions_<>() functions will have to be unit tested well, while modifying different default options and ensuring expected behaviour.
> 
> Unit tests should ensure all such combinations are covered. Please ensure that.
> 
> > > * Do you want to run the test with a default set of options? No -> What would you like to change? Exit -> Errors out. Would be good to go back to the default options question on Exit.
> > >   error_on_exit
> 
> I understand the use-case now Akkul. But the handling of this case should be looked into. It currently errors out when Exit is chosen from the menu. Maybe, catch the exception and log a good message.
> 
> > > * Interactive_mode.py, functions such as ask_questions_gui() and ask_questions_mobile() has be broken down. They are too long. Also, ensure this new module is pylint and radon compliant.
> 
> I do see that, you have broken down code well now. However, it has to be done a bit more. Pylint of this module is rated at 7.72/10 and radon run also points out few functions rated B/C which need to be addressed.

@PreedhiVivek ,
1) Removed line and changed _Note_ to note
2) removed unwanted print
3) I could not replicate the issue. But however, os_name, browser_versions, Os_version is used only when  running the test remotely. So i have made changes to the code such that you wont be able to select the OS Name or browser version unless your remote_flag is not set to "Yes". And I cant add more unit test scenarios as it requires human interaction with terminal for each test, and I gotta change the test file for every test case manually
4) Added message when exiting
5) Improved the code, now getting pyling rating of 9.05, now there are  four functions that show b,c rating in radon, i can't break those functions any smaller as they are nested if statements

**Comment by PreedhiVivek** [comment]
@akkuldn I am good with your changes.
One final formatting suggestion in Readme,
ReadMe_currently
To improve readability,
ReadMe_spacing

**Comment by akkuldn** [comment]
> @akkuldn I am good with your changes.
> One final formatting suggestion in Readme,
> ReadMe_currently
> To improve readability,
> ReadMe_spacing

Made the change @PreedhiVivek

**Pull request closed without merging**
